connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/executable/InboundExecutableRegistryImpl.java

```diff
@@ -0,0 +1,320 @@
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.inbound.executable;
+
+import io.camunda.connector.api.inbound.Health;
+import io.camunda.connector.api.inbound.ProcessElement;
+import io.camunda.connector.runtime.core.config.InboundConnectorConfiguration;
+import io.camunda.connector.runtime.core.inbound.InboundConnectorData;
+import io.camunda.connector.runtime.core.inbound.InboundConnectorElement;
+import io.camunda.connector.runtime.core.inbound.InboundConnectorFactory;
+import io.camunda.connector.runtime.inbound.executable.InboundExecutableEvent.Deactivated;
+import io.camunda.connector.runtime.inbound.executable.RegisteredExecutable.Activated;
+import io.camunda.connector.runtime.inbound.executable.RegisteredExecutable.ConnectorNotRegistered;
+import io.camunda.connector.runtime.inbound.executable.RegisteredExecutable.FailedToActivate;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Scheduled;
+
+public class InboundExecutableRegistryImpl implements InboundExecutableRegistry {
+
+  private final BlockingQueue<InboundExecutableEvent> eventQueue;
+  private final ExecutorService executorService;
+
+  private final BatchExecutableProcessor batchExecutableProcessor;
+
+  private final Map<ProcessElement, UUID> executablesByElement = new ConcurrentHashMap<>();
+  private final Map<UUID, RegisteredExecutable> executables = new HashMap<>();
+
+  private static final Logger LOG = LoggerFactory.getLogger(InboundExecutableRegistryImpl.class);
+
+  private final Map<String, List<String>> deduplicationScopesByType;
+
+  public InboundExecutableRegistryImpl(
+      InboundConnectorFactory connectorFactory, BatchExecutableProcessor batchExecutableProcessor) {
+    this.batchExecutableProcessor = batchExecutableProcessor;
+    this.executorService = Executors.newSingleThreadExecutor();
+    eventQueue = new LinkedBlockingQueue<>();
+    deduplicationScopesByType =
+        connectorFactory.getConfigurations().stream()
+            .collect(
+                Collectors.toMap(
+                    InboundConnectorConfiguration::type,
+                    InboundConnectorConfiguration::deduplicationProperties));
+    startEventProcessing();
+  }
+
+  void startEventProcessing() {
+    executorService.submit(
+        () -> {
+          try {
+            while (!Thread.currentThread().isInterrupted()) {
+              handleEvent(eventQueue.take()); // blocks until an event is available
```

Comment:
totally optional, but instead of smashing the `eventQueue` we could have a scheduled task every XXs (as the definitions are fetched every 5s anyway), and that task would process the events in batch. 
Again, this is just food for thoughts :)

Reply:
I'd just add that the thread will go to sleep until an element is available so this approach seems to be more efficient than polling